### PR TITLE
Introducing NetBSD userland

### PR DESCRIPTION
**Pull request by matthewbauer**
This adds some NetBSD userland into Nixpkgs. The main goal is to create some tools that will make adding more of NetBSD easier. The NetBSD tree is pulled from CVS directly and only needed paths are downloaded. FreeBSD and OpenBSD can use similar approaches.

This replaces the "musl-*" stuff that was really just hacked up versions of NetBSD sources. Hoping it doesn't break anything. It finally gave me an excuse to add my NetBSD work though.

/cc @dtzWill

### CODE REVIEW

**Review by Ericson2314 — APPROVED** [review]
Great job! You're really pushing us in a more portable direction and it's fantastic.

File: pkgs/top-level/unix-tools.nix [hunk 1]
```diff
@@ -44,6 +44,15 @@ in rec {
   eject = singleBinary "eject" {
     linux = pkgs.utillinux;
   };
+  getconf = singleBinary "getconf" {
+    linux = if (hostPlatform.libc == "glibc") then pkgs.glibc.bin
+                                              else pkgs.netbsd.getconf;
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/top-level/unix-tools.nix]: Alpine Linux uses netbsd sources. Didn’t know about busybox getent though.

[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/top-level/unix-tools.nix]: It looks like the busybox version is just a patched version of glibc’s.

File: pkgs/os-specific/bsd/netbsd/default.nix [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+
+    # hack for makefile check
+    SHLINKDIR="/usr/libexec";
+    SHLIBDIR="/usr/lib";
+
+    # not using bsd binutils
+    OBJCOPY = "echo fake objcopy";
+
+    # these values should be correct
+    # logic needs to change for cross compile
```
Comment (Ericson2314) [inline pkgs/os-specific/bsd/netbsd/default.nix]: It does? This seems correct to me :).

Comment (matthewbauer) [reply pkgs/os-specific/bsd/netbsd/default.nix]: Well these are used as the "target" by NetBSD (HOST_MACHINE_* should be set here but there are makefile issues)

Comment (Ericson2314) [reply pkgs/os-specific/bsd/netbsd/default.nix]: OK so the netbsd side is wack :grin:. Sounds good for now.

File: pkgs/top-level/unix-tools.nix [hunk 3]
```diff
@@ -49,12 +49,12 @@ in rec {
     linux = pkgs.utillinux;
   };
   getconf = singleBinary "getconf" {
-    linux = if hostPlatform.libc == "glibc" then lib.getBin pkgs.glibc
+    linux = if stdenv.cc.libc == "glibc" then lib.getBin pkgs.glibc
```
Comment (Ericson2314) [inline pkgs/top-level/unix-tools.nix]: `stdenv.cc.libc` is a derivation :D, you'll want to go back to how it was before.

**Comment by dtzWill** [comment]
yay!!

If you weren't already, can you make a PR with your unix-tools changes? I'd like to keep track of them and was planning to build on them once I have some time.  Mark `[WIP]` or w/e as appropriate, I'm not sure where they stand exactly.

Woohoo re:NetBSD!

**Comment by matthewbauer** [comment]
Yeah definitely. I just need to make sure it builds okay on Hydra before proposing any (potential) mass rebuilds.

**Comment by Ericson2314** [comment]
Just want to say great job! So much more maintainable than what Alpine linux does I gather.

**Comment by matthewbauer** [comment]
Hopefully! Still working my way through some of the initial failures though. ```netbsd.getent```, ```netbsd.fts``` and ```netbsd.getconf``` are working reliable though.

**Comment by dtzWill** [comment]
Tried switching to using these (still might for getent/getconf), but ran into troubles with `netbsd.fts` building `libselinux`:
* Seems to be misising includes, although documentation suggests `fts.h` expects code to include some things first.  Haven't investigated, because
* Seems to expect `__RENAME` macro from `<sys/cdefs.h>` which doesn't appear to be present in anyone's definition other than netbsd's own cdefs.

Thoughts?

**Comment by matthewbauer** [comment]
Yeah we need to add nbcompat as a propagated build input.

**Comment by dtzWill** [comment]
Hmm, looks like `getent` can't be built on musl either--due to `xinstall` needing `<fts.h>` via mtree headers. Hrmm.

```console
$ nix build -f . --arg localSystem '{config="x86_64-unknown-linux-musl";}' netbsd.install
error: build of '/nix/store/m11yvas0hcxzwsgpnzqxqv2lp3ri0xcl-xinstall-netbsd-7.1.2.drv' on 'ssh://wdietz2@presto' failed: builder for '/nix/store/m11yvas0hcxzwsgpnzqxqv2lp3ri0xcl-xinstall-netbsd-7.1.2.drv' failed with exit code 1
builder for '/nix/store/m11yvas0hcxzwsgpnzqxqv2lp3ri0xcl-xinstall-netbsd-7.1.2.drv' failed with exit code 1; last 10 log lines:
  In file included from /build/cvs-export/usr.sbin/mtree/getid.c:81:0:
  /build/cvs-export/usr.sbin/mtree/extern.h:43:10: fatal error: fts.h: No such file or directory
   #include <fts.h>
            ^~~~~~~
  compilation terminated.
  *** Error code 1
  
  Stop.
  make: stopped in /build/cvs-export/usr.bin/xinstall
  builder for '/nix/store/m11yvas0hcxzwsgpnzqxqv2lp3ri0xcl-xinstall-netbsd-7.1.2.drv' failed with exit code 1
[0 built (1 failed)]
error: build of '/nix/store/m11yvas0hcxzwsgpnzqxqv2lp3ri0xcl-xinstall-netbsd-7.1.2.drv' failed
```

If you want to poke at it, the above command should work as a template for testing...

**Comment by dtzWill** [comment]
Erm, I always forget the official binary cache doesn't have a musl toolchain.. yet O:).

Well I don't think it does.  Anyway, if that's a problem you can save some build time using something like:

```console
$ nix build --store $HOME/test-store --option binary-caches https://cache.allvm.org --option trusted-public-keys "gravity.cs.illinois.edu-1:yymmNS/WMf0iTj2NnD0nrVV8cBOXM9ivAkEdO1Lro3U= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=" -f ./nixpkgs-staging --arg localSystem '{config="x86_64-unknown-linux-musl";}' netbsd.getent
```

Which I'll note:
* Uses the ALLVM cache and key
* Uses local store so you don't have to root-level trust me/my builders
* As written expects your nixpkgs branch to be in `./nixpkgs-staging`--this very likely won't be right :).

**Comment by matthewbauer** [comment]
Ok I think it's fixed in https://github.com/matthewbauer/nixpkgs/tree/netbsd. Can you test it out before I push it to master?

**Comment by matthewbauer** [comment]
Those files shouldn't be installed. We just install what that include goes to. I think I have fixed your issue though. It's weird that I'm not getting it.

**Comment by matthewbauer** [comment]
@dtzWill Have you been able to compare musl-getent & netbsd.getent? I think they are close but depending on your use case may have some differences.

Anyway I would like to start looking into using some of the netbsd stuff in the wider nixpkgs. Has it been working properly for you since? I've committed a few different changes that I think have hit master by now.

**Pull request merged**